File: 2_module/Kruskal/Kruskal.cpp
#include <bits/stdc++.h>

using namespace std;

struct edge {
    int x, y;
    double dist;
};

double count_dist(pair <int, int> a, pair <int, int> b){
    return sqrt((a.first - b.first) * (a.first - b.first) + (a.second - b.second) * (a.second - b.second));
}

bool compare(edge a, edge b){
    return a.dist < b.dist;
}

int main(){
    int n;
    cin >> n;
    vector <pair <int, int > > points (n);
    for (int i = 0; i < n; i++){
        cin >> points[i].first >> points[i].second;
    }
    vector <edge> edges;
    for (int i = 0; i < n; i++){
        for (int j = i + 1; j < n; j++){
            edges.push_back({i, j, count_dist(points[i], points[j])});
            //cout << edges[i].x.first << " " << edges[j].x.second << edges[i].x.first << " " << edges[j].x.second << << " " << edges.dist << endl;
        }
    }
    sort(edges.begin(), edges.end(), compare);
    for (int i = 0; i < edges.size(); i++){
        cout << edges[i].dist << endl;
    }
    vector <int> tree_id(n);
    for (int i = 0; i < n; i++){
        tree_id[i] = i;
    }
    int m = edges.size();
    double cost = 0;
    for (int i = 0; i < m; i++){
        int a = edges[i].x, b = edges[i].y;
        double l = edges[i].dist;
        if (tree_id[a] != tree_id[b]){
            cost += l;
            int old_id = tree_id[b], new_id = tree_id[a];
            for (int j = 0; j < n; j++){
                if (tree_id[j] == old_id){
                    tree_id[j] = new_id;
                }
            }
        }
    }
    cout << cost;
}

File: 2_module/Kruskal/Kruskal_go/Kruskal.go
package main

import (
	"fmt"
	"math"
	"sort"
)

type edge struct {
	x, y int
	dist float64
}

type point struct {
	first, second int32
}

func count_dist(a, b point) float64 {
	return math.Sqrt(float64((a.first-b.first)*(a.first-b.first) + (a.second-b.second)*(a.second-b.second)))
}

func main() {
	var n int
	fmt.Scan(&n)
	points := make([]point, n, n)
	for i := 0; i < n; i++ {
		fmt.Scan(&points[i].first, &points[i].second)
	}
	edges := make([]edge, 0, 0)
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			edges = append(edges, edge{x: i, y: j, dist: count_dist(points[i], points[j])})
		}
	}
	sort.Slice(edges, func(i, j int) (less bool) {
		return edges[i].dist < edges[j].dist
	})
	tree_id := make([]int, n, n)
	for i := 0; i < n; i++ {
		tree_id[i] = i
	}
	m := len(edges)
	var cost float64 = 0
	for i := 0; i < m; i++ {
		a := edges[i].x
		b := edges[i].y
		l := edges[i].dist
		if tree_id[a] != tree_id[b] {
			cost += l
			old_id := tree_id[b]
			new_id := tree_id[a]
			for j := 0; j < n; j++ {
				if tree_id[j] == old_id {
					tree_id[j] = new_id
				}
			}
		}
	}
	fmt.Printf("%.2f", cost)
}
